parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java

```diff
@@ -0,0 +1,339 @@
+    private final BlockCipher.Encryptor metaDataEncryptor;
+    private final byte[] fileAAD ;
+
+    private byte[] dataPageHeaderAAD;
+    private byte[] dataPageAAD;
+    private byte[] dictPageHeaderAAD;
+    private byte[] dictPageAAD;
+
+    public EncryptorRunTime(InternalFileEncryptor fileEncryptor, ColumnChunkMetaData chunk,
+                            int blockId, int columnId) throws IOException  {
+      this.colEncrSetup = fileEncryptor.getColumnSetup(chunk.getPath(), true, toShortWithCheck(columnId));
+      this.dataEncryptor = colEncrSetup.getDataEncryptor();
+      this.metaDataEncryptor = colEncrSetup.getMetaDataEncryptor();
+
+      this.fileAAD = fileEncryptor.getFileAAD();
+      this.dataPageHeaderAAD = AesCipher.createModuleAAD(fileAAD, ModuleType.DataPageHeader,
+        toShortWithCheck(blockId), toShortWithCheck(columnId), (short) 0);
+      this.dataPageAAD = AesCipher.createModuleAAD(fileAAD, ModuleType.DataPage,
+        toShortWithCheck(blockId), toShortWithCheck(columnId), (short) 0);
+      this.dictPageHeaderAAD = AesCipher.createModuleAAD(fileAAD, ModuleType.DictionaryPageHeader,
+        toShortWithCheck(blockId), toShortWithCheck(columnId), (short) 0);
+      this.dictPageAAD = AesCipher.createModuleAAD(fileAAD, ModuleType.DictionaryPage,
+        toShortWithCheck(blockId), toShortWithCheck(columnId), (short) 0);
+    }
+
+    public InternalColumnEncryptionSetup getColEncrSetup() {
+      return this.colEncrSetup;
+    }
+
+    public BlockCipher.Encryptor getDataEncryptor() {
+      return this.dataEncryptor;
+    }
+
+    public BlockCipher.Encryptor getMetaDataEncryptor() {
+      return this.metaDataEncryptor;
+    }
+
+    public byte[] getFileAAD() {
+      return this.fileAAD;
+    }
+
+    public byte[] getDataPageHeaderAAD() {
+      return this.dataPageHeaderAAD;
+    }
+
+    public byte[] getDataPageAAD() {
+      return this.dataPageAAD;
+    }
+
+    public byte[] getDictPageHeaderAAD() {
+      return this.dictPageHeaderAAD;
+    }
+
+    public byte[] getDictPageAAD() {
+      return this.dictPageAAD;
+    }
+  }
+
+  private final int PAGE_BUFFER_SIZE = ParquetProperties.DEFAULT_PAGE_SIZE * 2;
+  private byte[] pageBuffer;
+  private Configuration conf;
+
+  public ColumnEncryptor(Configuration conf) {
+    this.pageBuffer = new byte[PAGE_BUFFER_SIZE];
+    this.conf = conf;
+  }
+
+  /**
+   * Given the input file, encrypt the columns specified by paths, and output the file.
+   * The encryption settings can be specified in the parameter of fileEncryptionProperties
+   * @param inputFile Input file
+   * @param outputFile Output file
+   * @param paths columns to be encrypted
+   * @param fileEncryptionProperties FileEncryptionProperties of the file
+   * @throws IOException
+   */
+  public void encryptColumns(String inputFile, String outputFile, List<String> paths, FileEncryptionProperties fileEncryptionProperties) throws IOException {
```

Comment:
an excellent point. I agree it would be good to support this scenario. It is somewhat harder to explain and define (check if each column is already encrypted; if yes, compare its encryption properties). It might be easier to start with a simpler scenario, where the input files are not encrypted at all. But I'm ok either way.

Reply:
please check if the file (or column) is already encrypted.